docs/writing-a-profile.md

```diff
@@ -0,0 +1,166 @@
+- The value for `product` relates to how a means of transport "works" *in local context*. Example: Even though [*S-Bahn*](https://en.wikipedia.org/wiki/Berlin_S-Bahn) and [*U-Bahn*](https://en.wikipedia.org/wiki/Berlin_U-Bahn) in Berlin are both `train`s, they have different operators, service patterns, stations and look different. Therefore, they are two distinct `product`s `subway` and `suburban`.
+
+**Specify `product`s that appear in the app** you recorded requests of. For a fictional transit network, this may look like this:
+
+```js
+const products = {
+	commuterTrain: {
+		product: 'commuterTrain',
+		mode: 'train',
+		bitmask: 1,
+		name: 'ACME Commuter Rail',
+		short: 'CR'
+	},
+	metro: {
+		product: 'metro',
+		mode: 'train',
+		bitmask: 2,
+		name: 'Foo Bar Metro',
+		short: 'M'
+	}
+}
+```
+
+Let's break this down:
+
+- `product`: A sensible, [camelCased](https://en.wikipedia.org/wiki/Camel_case#Variations_and_synonyms), alphanumeric identifier. Use it for the key in the `products` object as well.
+- `mode`: A [valid *Friendly Public Transport Format* `1.0.1` mode](https://github.com/public-transport/friendly-public-transport-format/blob/1.0.1/spec/readme.md#modes).
+- `bitmask`: HAFAS endpoints work with a [bitmask](https://en.wikipedia.org/wiki/Mask_(computing)#Arguments_to_functions) that toggles the individual products. the value should toggle the appropriate bit(s) in the bitmask (see below).
+- `name`: A short, but distinct name for the means of transport, *just precise enough in local context*. In Berlin, `S-Bahn commuter rail` would be too much, because everyone knows what `S-Bahn` means.
+- `short`: The shortest possible symbol that identifies the product.
+
+todo: `defaultProducts`, `allProducts`, `bitmasks`, add to profile
+
+If you want, you can now **verify that the profile works**; I've prepared [a script](https://runkit.com/derhuerst/hafas-client-profile-example) for that. Alternatively, [submit a Pull Request](https://help.github.com/articles/creating-a-pull-request-from-a-fork/) and I will help you out with testing and improvements.
+
+### Finding the right values for the `bitmask` field
+
+As shown in [the video](https://stuff.jannisr.de/how-to-record-hafas-requests.mp4), search for a journey and toggle off one product at a time, recording the requests. After extracting the products bitmask ([example](https://gist.github.com/derhuerst/193ef489f8aa50c2343f8bf1f2a22069#file-via-http-L34)) you will end up with values looking like these:
+
+```
+toggles                     value  binary  subtraction     bit(s)
+all products                31     11111   31 - 0
+all but ACME Commuter Rail  15     01111   31 - 2^4        2^4
+all but Foo Bar Metro       23     10111   31 - 2^3        2^3
+all but product E           30     11001   31 - 2^2 - 2^1  2^2, 2^1
+all but product F           253    11110   31 - 2^1        2^0
+```
```

Comment:
It might actually be even easier to understand if you don't do 'all but product x' but rather 'only product x' since that directly gives you the corresponding powers of 2. Especially easier to understand if one product has multiple values (like for ÖBB).

Reply:
Then it would be misaligned with the tutorial video though. Do you think it would still be more intuitive? Is it hard to understand right now?

Reply:
Ah, I didn't watch the video. Keep it like this then, it's fine enough.